skfp/descriptors/constitutional.py

```diff
@@ -0,0 +1,252 @@
+from typing import Optional, Union
+
+from rdkit.Chem import AddHs, Mol
+from rdkit.Chem.Descriptors import MolWt
+from rdkit.Chem.rdMolDescriptors import CalcNumRings, CalcNumRotatableBonds
+
+
+def validate_molecule(func):
+    """
+    Decorator to validate that the molecule has at least one atom.
+
+    Raises
+    ------
+    ValueError
+        If the molecule has no atoms.
+    """
+
+    def wrapper(mol: Mol, *args, **kwargs):
+        if mol.GetNumAtoms() == 0:
+            raise ValueError(
+                f"The molecule has no atoms, {func.__name__} cannot be calculated."
+            )
+        return func(mol, *args, **kwargs)
+
+    return wrapper
+
+
+@validate_molecule
+def average_molecular_weight(mol: Mol) -> float:
+    """
+    Average Molecular Weight.
+
+    Calculates the average molecular weight of the molecule, defined as the molecular
+    weight divided by the number of atoms.
+
+    This is different from "average molecular weight" in the context of isotopes [1]_.
+
+    Parameters
+    ----------
+    mol : RDKit Mol object
+        The molecule for which the average molecular weight is to be calculated.
+
+    References
+    ----------
+    .. [1] `
+        <https://chemistry.stackexchange.com/questions/150993/discrepancy-when-calulating-mol-weights-with-chemsketch-and-python-rdkit/150999#150999>`_
+
+    Examples
+    --------
+    >>> from rdkit.Chem import MolFromSmiles
+    >>> from skfp.descriptors.constitutional import average_molecular_weight
+    >>> mol = MolFromSmiles("C1=CC=CC=C1")  # Benzene
+    >>> average_molecular_weight(mol)
+    13.019
+    """
+    return MolWt(mol) / mol.GetNumAtoms()
+
+
+@validate_molecule
+def bond_type_count(mol: Mol, bond_type: Optional[str] = None) -> int:
+    """
+    Bond Type Count.
+
+    Counts the total number of bonds of a specific type in the molecule.
+
+    Parameters
+    ----------
+    mol : RDKit Mol object
+        The molecule for which the bond count is to be calculated.
+
+    bond_type : str, optional
+        The type of bond to count. Valid options are:
+        - "SINGLE"
+        - "DOUBLE"
+        - "TRIPLE"
+        - "AROMATIC"
+        If None, the function returns the total number of bonds.
+
+    Examples
+    --------
+    >>> from rdkit.Chem import MolFromSmiles
+    >>> mol = MolFromSmiles("C1=CC=CC=C1")  # Benzene
+    >>> bond_type_count(mol, "AROMATIC")
+    6
+    >>> bond_type_count(mol, "DOUBLE")
+    0
+    >>> bond_type_count(mol)  # Total bonds
+    6
+    """
+    if bond_type:
+        return sum(
+            1 for bond in mol.GetBonds() if bond.GetBondType().name == bond_type.upper()
+        )
+    return mol.GetNumBonds()
+
+
+@validate_molecule
+def element_atom_count(mol: Mol, atom_id: Union[int, str]) -> int:
+    """
+    Element atom count.
+
+    Calculates the count of atoms of a specific type.
+
+    The function returns the total number of Hs (explicit and implicit) on the atom.
```

Comment:
1. No empty line in between
2. Maybe be slightly more specific with the second sentence, e.g. `In case of hydrogens, the total number is returned, counting both explicit and implicit ones`.